drivers/usb/udc/udc_common.c

```diff
@@ -0,0 +1,1086 @@
+					   struct net_buf *const setup,
+					   const uint8_t ep)
+{
+	size_t size = udc_data_stage_length(setup);
+	struct udc_buf_info *bi;
+	struct net_buf *buf;
+
+	buf = udc_ctrl_alloc_stage(dev, setup, ep, size);
+	if (buf) {
+		bi = udc_get_buf_info(buf);
+		bi->data = true;
+	}
+
+	return buf;
+}
+
+static struct net_buf *udc_ctrl_alloc_status(const struct device *dev,
+					     struct net_buf *const parent,
+					     const uint8_t ep)
+{
+	size_t size = (ep == USB_CONTROL_EP_OUT) ? 64 : 0;
+	struct udc_buf_info *bi;
+	struct net_buf *buf;
+
+	buf = udc_ctrl_alloc_stage(dev, parent, ep, size);
+	if (buf) {
+		bi = udc_get_buf_info(buf);
+		bi->status = true;
+	}
+
+	return buf;
+}
+
+int udc_ctrl_submit_s_out_ack(const struct device *dev,
+			      struct net_buf *const dout)
+{
+	struct udc_buf_info *bi = udc_get_buf_info(dout);
+	struct udc_data *data = dev->data;
+	struct net_buf *buf;
+	int ret = 0;
+
+	bi->data = true;
+	net_buf_frag_add(data->setup, dout);
+
+	buf = udc_ctrl_alloc_status(dev, dout, USB_CONTROL_EP_IN);
+	if (buf == NULL) {
+		ret = -ENOMEM;
+	}
+
+	return udc_submit_event(dev, UDC_EVT_EP_REQUEST, ret, data->setup);
+}
+
+int udc_ctrl_submit_s_in_ack(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+	struct net_buf *buf;
+	int ret = 0;
+
+	if (!udc_ctrl_stage_is_data_in(dev)) {
+		return -ENOTSUP;
+	}
+
+	/* Allocate buffer for data stage IN */
+	buf = udc_ctrl_alloc_data(dev, data->setup, USB_CONTROL_EP_IN);
+	if (buf == NULL) {
+		ret = -ENOMEM;
+	}
+
+	return udc_submit_event(dev, UDC_EVT_EP_REQUEST, ret, data->setup);
+}
+
+int udc_ctrl_submit_s_ack(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+	struct net_buf *buf;
+	int ret = 0;
+
+	/* Allocate buffer for possible status IN */
+	buf = udc_ctrl_alloc_status(dev, data->setup, USB_CONTROL_EP_IN);
+	if (buf == NULL) {
+		ret = -ENOMEM;
+	}
+
+	return udc_submit_event(dev, UDC_EVT_EP_REQUEST, ret, data->setup);
+}
+
+int udc_ctrl_submit_status(const struct device *dev,
+			   struct net_buf *const buf)
+{
+	struct udc_buf_info *bi = udc_get_buf_info(buf);
+
+	bi->status = true;
+
+	return udc_submit_event(dev, UDC_EVT_EP_REQUEST, 0, buf);
+}
+
+bool udc_ctrl_stage_is_data_out(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_DATA_OUT ? true : false;
+}
+
+bool udc_ctrl_stage_is_data_in(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_DATA_IN ? true : false;
+}
+
+bool udc_ctrl_stage_is_status_out(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_STATUS_OUT ? true : false;
+}
+
+bool udc_ctrl_stage_is_status_in(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_STATUS_IN ? true : false;
+}
+
+bool udc_ctrl_stage_is_no_data(const struct device *dev)
```

Comment:
> There is no bus traffic if SETUP wLength is 0.

What is your statement referring to.

Reply:
Consider for example SET CONFIGURATION request to device with address 3.

The SETUP stage consists of single transaction:
```
Host->Device | 2d 03 50                         | SETUP Address 3 Endpoint 0 CRC5: 0x0a
Host->Device | c3 00 09 01 00 00 00 00 00 27 25 | DATA0 (0xc3) Actual Request (8 bytes) CRC16 (0x2527)
Device->Host | d2                               | ACK
```

Because wLength was 0, there is no Data stage and there aren't any transactions on the bus. The state machine directly goes into Status stage.

The Status stage can contain multiple NAKed IN transactions:
```
Host->Device | 69 03 50 | IN Address 3 Endpoint 0 CRC5: 0x0a
Device->Host | 5a       | NAK
```

The Status stage ends with acknowledged IN transaction:
```
Host->Device | 69 03 50 | IN Address 3 Endpoint 0 CRC5: 0x0a
Device->Host | 4b 00 00 | DATA1 (ZLP) CRC: 0x0000
Host->Device | d2       | ACK
```